lib/asciidoctor/syntax_highlighter/coderay.rb

```diff
@@ -0,0 +1,90 @@
+module Asciidoctor
+  class SyntaxHighlighter::CodeRay < SyntaxHighlighter::Base
+    register_for 'coderay'
+
+    def initialize *args
+      super
+      @requires_stylesheet = nil
+    end
+
+    def highlight?
+      coderay_available?
+    end
+
+    # NOTE CodeRay::Duo gracefully falls back to no highlighting if lang isn't recognized
+    def highlight node, source, lang, opts
+      @requires_stylesheet = true if (css_mode = opts[:css_mode]) == :class
+      highlighted = ::CodeRay::Duo[(lang || :text).to_sym, :html, {
+        css: css_mode,
+        line_numbers: (line_numbers = opts[:line_numbers]),
+        line_number_start: opts[:start_line_number],
+        line_number_anchors: false,
+        highlight_lines: opts[:highlight_lines],
+        bold_every: false,
+      }].highlight source
+      if line_numbers == :table && opts[:callouts]
+        [highlighted, (idx = highlighted.index CodeCellStartTagCs) ? idx + CodeCellStartTagCs.length : nil]
+      else
+        highlighted
+      end
+    end
+
+    def format node, lang, opts
+      code_attrs = lang ? %( data-lang="#{lang}") : ''
+      %(<pre class="CodeRay highlight#{opts[:nowrap] ? ' nowrap' : ''}"><code#{code_attrs}>#{node.content}</code></pre>)
+    end
+
+    def docinfo? location
+      @requires_stylesheet && location == :footer
+    end
+
+    def docinfo location, doc, opts
+      if opts[:linkcss]
+        %(<link rel="stylesheet" href="#{doc.normalize_web_path stylesheet_filename, (doc.attr 'stylesdir', ''), false}"#{opts[:self_closing_tag_slash]}>)
+      else
+        %(<style>
+#{stylesheet}
+</style>)
+      end
+    end
+
+    def write_stylesheet? doc
+      @requires_stylesheet
+    end
+
+    def write_stylesheet doc, to_dir
+      ::File.write (::File.join to_dir, stylesheet_filename), stylesheet
+    end
+
+    module Loader
+      private
+
+      def coderay_available?
+        (@@status ||= load_coderay) == :loaded
+      end
+
+      def load_coderay
+        (defined? ::CodeRay::Duo) ? :loaded : (Helpers.require_library 'coderay', true, :warn).nil? ? :unavailable : :loaded
+      end
+    end
+
+    module Styles
+      include Loader
+
+      def stylesheet
```

Comment:
@Mogztter Do you like the name `stylesheet` or `stylesheet_data` better? (or something else?)

Reply:
I like these suggestions, especially stylesheet_basename.

I'm still torn between read_stylesheet and load_stylesheet. However, I'm leading more towards "read" because "load" implies it's either being parsed into an object model or loaded into another object. So I'll start with read_stylesheet and we'll see how it feels.